README
= escape - HTML/URI/shell escaping utilities


escape library provides several HTML/URI/shell escaping functions.

== Author

Tanaka Akira <akr@fsij.org>

== Home Page

http://www.a-k-r.org/escape/

== Feature

* several escaping/composing functions
  * HTML text
  * HTML attribute value
  * HTML form (x-www-form-urlencoded)
  * URI path
  * shell command line
* dedicated classes for escaped strings
* escape and compose strongly related strings at once

== Usage

  require 'escape'

  Escape.shell_command(["echo", "*"]) #=> #<Escape::ShellEscaped: echo '*'>
  Escape.uri_path("a?b/c?d/e?f") #=> #<Escape::PercentEncoded: a%3Fb/c%3Fd/e%3Ff>
  Escape.html_form([["a","b"], ["c","d"]]) #=> #<Escape::PercentEncoded: a=b&c=d>
  Escape.html_form({"a"=>"b", "c"=>"d"}) #=> #<Escape::PercentEncoded: a=b&c=d>
  Escape.html_text("a & b < c > d") #=> #<Escape::HTMLEscaped: a &amp; b &lt; c &gt; d>
  Escape.html_attr_value("ab&<>\"c") #=> #<Escape::HTMLAttrValue: "ab&amp;&lt;&gt;&quot;c">

== Requirements

* ruby : http://www.ruby-lang.org/

== Download

* latest release: http://www.a-k-r.org/escape/escape-0.2.tar.gz

* development version in Subversion repository:

    % svn co svn://svn@svn.a-k-r.org/akr/escape/trunk escape

== Install

  % ruby install.rb

== Reference Manual

See rdoc/classes/Escape.html or
http://www.a-k-r.org/escape/rdoc/classes/Escape.html

== License

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

(1) Redistributions of source code must retain the above copyright notice, this
    list of conditions and the following disclaimer.
(2) Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.
(3) The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
OF SUCH DAMAGE.

(The modified BSD licence)


ChangeLog
2009-10-04  Tanaka Akira  <akr@fsij.org>

	* escape.rb (Escape.percent_encoding): remove duplicates in regex
	  character range to avoid a warning.

2007-06-03  Tanaka Akira  <akr@fsij.org>

	* escape.rb (Escape::MIMEParameter): new class.
	  (Escape.mime_token?): new method.
	  (Escape.rfc2822_quoted_string): ditto.
	  (Escape.mime_parameter_value): ditto.
	  (Escape.mime_parameter): ditto.
	  (Escape.http_token?): ditto.
	  (Escape.http_quoted_string): ditto.
	  (Escape.http_parameter_value): ditto.
	  (Escape.http_parameter): ditto.
	  (Escape.http_params_with_sep): ditto.
	  (Escape.http_params_with_pre): ditto.

2007-03-05  Tanaka Akira  <akr@fsij.org>

	* escape.rb (Escape::StringWrapper#escaped_string): New method.

2007-03-03  Tanaka Akira  <akr@fsij.org>

	* escape.rb: Escape.uri_path accepts an array of strings.

	* escape.rb: Escape::PercentEncoded#split_html_form implemented.

	* escape.rb: Escape.percent_encoding implemented.

2007-02-28  Tanaka Akira  <akr@fsij.org>

	* version 0.2 released.

2007-02-28  Tanaka Akira  <akr@fsij.org>

	* escape.rb: Escape::HTMLAttrValue defined.
	  Escape.html_attr_value returns it.

2007-02-27  Tanaka Akira  <akr@fsij.org>

	* escape.rb: make StringWrapper into class.
	  Escape::ShellEscaped, Escape::HTMLEscaped and Escape::PercentEncoded
	  inherit it.

2007-02-26  Tanaka Akira  <akr@fsij.org>

	* escape.rb: Escape::ShellEscaped defined.
	  Escape.shell_command and Escape.shell_single_word returns
	  Escape::ShellEscaped object instead of a string.

	* escape.rb: Escape::HTMLEscaped defined.
	  Escape.html_text and Escape.html_attr_value returns
	  Escape::HTMLEscaped object instead of a string.

	* escape.rb: Escape::PercentEncoded defined.
	  Escape.uri_segment, Escape.uri_path and Escape.html_form returns
	  Escape::PercentEncoded object instead of a string.

	* escape.rb: rename Escape.html_attr to Escape.html_attr_value.

2007-01-05  Tanaka Akira  <akr@fsij.org>

	* version 0.1 released.


test/test-escape.rb
require 'test/unit'
require 'escape'

class TestEscapeStringWrapper < Test::Unit::TestCase
  def test_eq
    assert(Escape::PercentEncoded.new("foo") == Escape::PercentEncoded.new("foo"))
    assert(Escape::PercentEncoded.new("foo") != Escape::PercentEncoded.new("bar"))
    assert(Escape::ShellEscaped.new("a") != Escape::PercentEncoded.new("a"))
  end

  def test_hash
    v1 = Escape::PercentEncoded.new("foo")
    v2 = Escape::PercentEncoded.new("foo")
    h = {}
    h[v1] = 1
    h[v2] = 2
    assert_equal(1, h.size)
    assert_equal(2, h[v1])
  end

  def test_new_dup
    s = "a"
    o = Escape::PercentEncoded.new(s)
    assert_not_equal(s.object_id, o.instance_variable_get(:@str).object_id)
    o = Escape::PercentEncoded.new_no_dup(s)
    assert_equal(s.object_id, o.instance_variable_get(:@str).object_id)
  end

  def test_escaped_string
    assert_equal("a", Escape::PercentEncoded.new("a").escaped_string)
  end
end

class TestEscapeShellEscaped < Test::Unit::TestCase
  def assert_equal_se(str, tst)
    assert_equal(Escape::ShellEscaped.new(str), tst)
  end

  def test_shell_command
    assert_equal_se("com arg", Escape.shell_command(%w[com arg]))
    assert_equal_se("ls /", Escape.shell_command(%w[ls /]))
    assert_equal_se("echo '*'", Escape.shell_command(%w[echo *]))
  end

  def test_shell_single_word
    assert_equal_se("''", Escape.shell_single_word(''))
    assert_equal_se("foo", Escape.shell_single_word('foo'))
    assert_equal_se("'*'", Escape.shell_single_word('*'))
  end

end

class TestEscapePercentEncoded < Test::Unit::TestCase
  def str_to_pe(obj)
    case obj
    when String
      Escape::PercentEncoded.new(obj)
    when Array
      obj.map {|e| str_to_pe(e) }
    else
      obj
    end
  end

  def assert_equal_pe(str, tst)
    assert_equal(str_to_pe(str), tst)
  end

  def test_percent_encoding
    assert_equal_pe(
      "%20%21%22%23%24%25%26%27%28%29%2A%2B%2C-.%2F"+
      "%3A%3B%3C%3D%3E%3F%40%5B%5C%5D%5E_%60%7B%7C%7D~",
      Escape.percent_encoding(' !"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'))
  end

  def test_uri_segment
    assert_equal_pe("a%2Fb", Escape.uri_segment("a/b"))
  end

  def test_uri_path
    assert_equal_pe("a/b/c", Escape.uri_path("a/b/c"))
    assert_equal_pe("a%3Fb/c%3Fd/e%3Ff", Escape.uri_path("a?b/c?d/e?f"))
  end

  def test_uri_path_ary
    assert_equal_pe("a/b/c", Escape.uri_path(%w[a b c]))
    assert_equal_pe("%2Fd/f", Escape.uri_path(%w[/d f]))
  end

  def test_html_form
    assert_kind_of(Escape::PercentEncoded, Escape.html_form([["foo","bar"]]))
    assert_equal_pe("a=b&c=d", Escape.html_form([["a","b"], ["c","d"]]))
    assert_equal_pe("a=b;c=d", Escape.html_form([["a","b"], ["c","d"]], ';'))
    assert_equal_pe("k=1&k=2", Escape.html_form([["k","1"], ["k","2"]]))
    assert_equal_pe("k%3D=%26%3B%3D", Escape.html_form([["k=","&;="]]))
  end

  def test_split_html_form
    assert_equal_pe([["a", "b"], ["c", "d"]], Escape::PercentEncoded.new("a=b&c=d").split_html_form)
    assert_equal_pe([["a", "b"], ["c", "d"]], Escape::PercentEncoded.new("a=b;c=d").split_html_form)
    assert_equal_pe([["k", "1"], ["k", "2"]], Escape::PercentEncoded.new("k=1&k=2").split_html_form)
    assert_equal_pe([["k%3D", "%26%3B%3D"]], Escape::PercentEncoded.new("k%3D=%26%3B%3D").split_html_form)
    assert_raise(Escape::InvalidHTMLForm) { Escape::PercentEncoded.new("a=b;cd").split_html_form }
    assert_raise(Escape::InvalidHTMLForm) { Escape::PercentEncoded.new("a=b;c=d;").split_html_form }
  end
end

class TestEscapeHTML < Test::Unit::TestCase
  def assert_equal_he(str, tst)
    assert_equal(Escape::HTMLEscaped.new(str), tst)
  end

  def assert_equal_hav(str, tst)
    assert_equal(Escape::HTMLAttrValue.new(str), tst)
  end

  def test_html_text
    assert_equal_he('a&amp;&lt;&gt;"', Escape.html_text('a&<>"'))
  end

  def test_html_attr_value
    assert_equal_hav('"a&amp;&lt;&gt;&quot;"', Escape.html_attr_value('a&<>"'))
  end
end

class TestEscapeMIME < Test::Unit::TestCase
  def assert_equal_mime(str, tst)
    assert_equal(Escape::MIMEParameter.new(str), tst)
  end

  US_ASCII = (0x00..0x7f).map {|byte| [byte].pack("C") }
  CTLs = (0x00..0x1f).map {|byte| [byte].pack("C") } + ["\x7f"]
  SPACE = (0x00..0x1f).map {|byte| [byte].pack("C") } + ["\x7f"]
  OCTETS = (0x00..0xff).map {|byte| [byte].pack("C") }

  MIME_TSPECIALS = [
    "(", ")", "<", ">", "@",
    ",", ";", ":", "\\", '"',
    "/", "[", "]", "?", "="
  ]

  def test_mime_token
    assert_equal(true, Escape.mime_token?("abc"))
    assert_equal(false, Escape.mime_token?("a=b"))
    MIME_TSPECIALS.each {|ch|
      assert_equal(false, Escape.mime_token?(ch))
    }
    (US_ASCII-[" "]-CTLs-MIME_TSPECIALS).each {|ch|
      assert_equal(true, Escape.mime_token?(ch))
    }
  end

  def test_mime_parameter_value
    assert_equal_mime('abc', Escape.mime_parameter_value("abc"))
    assert_equal_mime('"a/b/c"', Escape.mime_parameter_value("a/b/c"))
    assert_equal_mime('"\""', Escape.mime_parameter_value('"'))
    assert_equal_mime("\"a\n b\"", Escape.mime_parameter_value("a\n b"))
    assert_equal_mime("\"a\n\tb\"", Escape.mime_parameter_value("a\n\tb"))
    assert_equal_mime("\"a\r\n b\"", Escape.mime_parameter_value("a\r\n b"))
    assert_equal_mime("\"a\r\n\tb\"", Escape.mime_parameter_value("a\r\n\tb"))
    assert_equal_mime("\"\n \"", Escape.mime_parameter_value("\n "))
    assert_raise(ArgumentError) { Escape.mime_parameter_value("\n \n ") }
    assert_raise(ArgumentError) { Escape.mime_parameter_value("\r") }
    assert_raise(ArgumentError) { Escape.mime_parameter_value("\n") }
    assert_raise(ArgumentError) { Escape.mime_parameter_value("\0") }
  end

  def test_mime_parameter
    assert_equal_mime('abc=def', Escape.mime_parameter("abc", "def"))
    assert_equal_mime('aa="a/b/c"', Escape.mime_parameter("aa", "a/b/c"))
    assert_raise(ArgumentError) { Escape.mime_parameter("/", "\r") }
    assert_raise(ArgumentError) { Escape.mime_parameter("n", "\r") }
    assert_raise(ArgumentError) { Escape.mime_parameter("n", "\n") }
    assert_raise(ArgumentError) { Escape.mime_parameter("n", "\0") }
  end

  HTTP_SEPARATORS = [
    "(", ")", "<", ">", "@",
    ",", ";", ":", "\\", '"',
    "/", "[", "]", "?", "=",
    "{", "}", " ", "\t"
  ]

  def test_http_token
    assert_equal(true, Escape.http_token?("abc"))
    assert_equal(false, Escape.http_token?("a=b"))
    HTTP_SEPARATORS.each {|ch|
      assert_equal(false, Escape.http_token?(ch))
    }
    (US_ASCII-CTLs-HTTP_SEPARATORS).each {|ch|
      assert_equal(true, Escape.http_token?(ch))
    }
  end

  def test_http_quoted_string
    assert_equal_mime('"abc"', Escape.http_quoted_string("abc"))
    assert_equal_mime('"a/b/c"', Escape.http_parameter_value("a/b/c"))
    assert_equal_mime('"\""', Escape.http_parameter_value('"'))
    assert_equal_mime('"\\\\"', Escape.http_parameter_value('\\'))
    assert_equal_mime("\"a\n b\"", Escape.http_parameter_value("a\n b"))
    assert_equal_mime("\"a\n\tb\"", Escape.http_parameter_value("a\n\tb"))
    assert_equal_mime("\"a\r\n b\"", Escape.http_parameter_value("a\r\n b"))
    assert_equal_mime("\"a\r\n\tb\"", Escape.http_parameter_value("a\r\n\tb"))
    assert_equal_mime("\"\n \n \"", Escape.http_parameter_value("\n \n "))
    assert_raise(ArgumentError) { Escape.http_quoted_string("\n") }

  end

  def test_http_params_with_sep
    assert_equal_mime('', Escape.http_params_with_sep("; "))
    assert_equal_mime('a=b', Escape.http_params_with_sep("; ", "a", "b"))
    assert_equal_mime('a=b; c=d', Escape.http_params_with_sep("; ", "a", "b", "c", "d"))
    assert_raise(ArgumentError) { Escape.http_params_with_sep(";", "x") }
  end

  def test_http_params_with_pre
    assert_equal_mime('', Escape.http_params_with_pre("; "))
    assert_equal_mime('; a=b', Escape.http_params_with_pre("; ", "a", "b"))
    assert_equal_mime('; a=b; c=d', Escape.http_params_with_pre("; ", "a", "b", "c", "d"))
    assert_raise(ArgumentError) { Escape.http_params_with_pre(";", "x") }
  end
end


install.rb
#!/usr/bin/env ruby

# usage: ruby install.rb [-n] [--destdir=DESTDIR] [--vendor]
# options:
#  -n : don't install
#  --destdir=DESTDIR
#  --vendor
#
# Author: Tanaka Akira <akr@m17n.org>

require 'optparse'
require 'fileutils'

def target_directory(vendor)
  pat = vendor ? %r{/vendor_ruby/[\d.]+\z} : %r{/site_ruby/[\d.]+\z}
  $:.each {|loc|
    if pat =~ loc
      return loc
    end
  }
  raise "could not find target install directory"
end

CVS_FILES = {}
def cvs_files(dir)
  return CVS_FILES[dir] if CVS_FILES.include? dir
  if File.directory? "#{dir}/CVS"
    result = {}
    File.foreach("#{dir}/CVS/Entries") {|line|
      case line
      when %r{\A/([^/]+)/} then result[$1] = true
      when %r{\AD/([^/]+)/} then result[$1] = true
      end
    }
  else
    result = nil
  end
  CVS_FILES[dir] = result
  result
end

def each_target(&block)
  target_set = {}
  cvs = cvs_files('.')
  Dir.glob("*.rb") {|filename|
    next if /\Atest-/ =~ filename
    next if /\Ainstall/ =~ filename
    next if cvs && !cvs.include?(filename)
    target_set[filename] = true
    yield filename
    each_require(filename, target_set, &block)
  }
end

def each_require(file, target_set, &block)
  File.foreach(file) {|line|
    next if /\A\s*require\s+['"]([^'"]+)['"]/ !~ line
    feature = $1
    filename = "#{feature}.rb"
    next if target_set.include? filename
    next if !File.exist?(filename)
    target_set[filename] = true
    yield filename
    each_require(filename, target_set, &block)
  }
end

def collect_target
  result = []
  each_target {|filename| result << filename }
  result.sort!
  result
end

def install_file(src, dst)
  ignore_exc(Errno::ENOENT) { return if FileUtils.compare_file src, dst }
  # check shadow
  ignore_exc(Errno::ENOENT) { File.unlink dst }
  FileUtils.mkdir_p(File.dirname(dst), :mode=>0755)
  FileUtils.cp(src, dst, :verbose => true)
  File.chmod(0644, dst)
end

def ignore_exc(exc)
  begin
    yield
  rescue exc
  end
end

$opt_n = false
$opt_destdir = ""
$opt_vendor = false
ARGV.options {|q|
  q.banner = 'ruby install.rb [opts]'
  q.def_option('--help', 'show this message') {puts q; exit(0)}
  q.def_option('-n', "don't install") { $opt_n = true }
  q.def_option('--destdir=DESTDIR', "specify DESTDIR") {|destdir| $opt_destdir = destdir }
  q.def_option('--vendor', 'install to vendor directory') { $opt_vendor = true }
  q.parse!
}

if $opt_n
  dir = target_directory($opt_vendor)
  collect_target.each {|filename|
    puts "-> #{$opt_destdir}#{dir}/#{filename}"
  }
  exit
else
  File.umask 022
  dir = target_directory($opt_vendor)
  collect_target.each {|filename|
    install_file filename, "#{$opt_destdir}#{dir}/#{filename}"
  }
end



escape.rb
# escape.rb - escape/unescape library for several formats
#
# Copyright (C) 2006,2007 Tanaka Akira  <akr@fsij.org>
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
#  1. Redistributions of source code must retain the above copyright notice, this
#     list of conditions and the following disclaimer.
#  2. Redistributions in binary form must reproduce the above copyright notice,
#     this list of conditions and the following disclaimer in the documentation
#     and/or other materials provided with the distribution.
#  3. The name of the author may not be used to endorse or promote products
#     derived from this software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.

# Escape module provides several escape functions.
# * URI
# * HTML
# * shell command
# * MIME parameter
module Escape
  module_function

  class StringWrapper
    class << self
      alias new_no_dup new
      def new(str)
        new_no_dup(str.dup)
      end
    end

    def initialize(str)
      @str = str
    end

    def escaped_string
      @str.dup
    end

    alias to_s escaped_string

    def inspect
      "\#<#{self.class}: #{@str}>"
    end

    def ==(other)
      other.class == self.class && @str == other.instance_variable_get(:@str)
    end
    alias eql? ==

    def hash
      @str.hash
    end
  end

  class ShellEscaped < StringWrapper
  end

  # Escape.shell_command composes
  # a sequence of words to
  # a single shell command line.
  # All shell meta characters are quoted and
  # the words are concatenated with interleaving space.
  # It returns an instance of ShellEscaped.
  #
  #  Escape.shell_command(["ls", "/"]) #=> #<Escape::ShellEscaped: ls />
  #  Escape.shell_command(["echo", "*"]) #=> #<Escape::ShellEscaped: echo '*'>
  #
  # Note that system(*command) and
  # system(Escape.shell_command(command).to_s) is roughly same.
  # There are two exception as follows.
  # * The first is that the later may invokes /bin/sh.
  # * The second is an interpretation of an array with only one element: 
  #   the element is parsed by the shell with the former but
  #   it is recognized as single word with the later.
  #   For example, system(*["echo foo"]) invokes echo command with an argument "foo".
  #   But system(Escape.shell_command(["echo foo"]).to_s) invokes "echo foo" command
  #   without arguments (and it probably fails).
  def shell_command(command)
    s = command.map {|word| shell_single_word(word) }.join(' ')
    ShellEscaped.new_no_dup(s)
  end

  # Escape.shell_single_word quotes shell meta characters.
  # It returns an instance of ShellEscaped.
  #
  # The result string is always single shell word, even if
  # the argument is "".
  # Escape.shell_single_word("") returns #<Escape::ShellEscaped: ''>.
  #
  #  Escape.shell_single_word("") #=> #<Escape::ShellEscaped: ''>
  #  Escape.shell_single_word("foo") #=> #<Escape::ShellEscaped: foo>
  #  Escape.shell_single_word("*") #=> #<Escape::ShellEscaped: '*'>
  def shell_single_word(str)
    if str.empty?
      ShellEscaped.new_no_dup("''")
    elsif %r{\A[0-9A-Za-z+,./:=@_-]+\z} =~ str
      ShellEscaped.new(str)
    else
      result = ''
      str.scan(/('+)|[^']+/) {
        if $1
          result << %q{\'} * $1.length
        else
          result << "'#{$&}'"
        end
      }
      ShellEscaped.new_no_dup(result)
    end
  end

  class InvalidHTMLForm < StandardError
  end
  class PercentEncoded < StringWrapper
    # Escape::PercentEncoded#split_html_form decodes
    # percent-encoded string as
    # application/x-www-form-urlencoded
    # defined by HTML specification.
    #
    # It recognizes "&" and ";" as a separator of key-value pairs.
    #
    # If it find is not valid as
    # application/x-www-form-urlencoded,
    # Escape::InvalidHTMLForm exception is raised.
    #
    #  Escape::PercentEncoded.new("a=b&c=d")
    #  #=> [[#<Escape::PercentEncoded: a>, #<Escape::PercentEncoded: b>],
    #       [#<Escape::PercentEncoded: c>, #<Escape::PercentEncoded: d>]]
    #
    #  Escape::PercentEncoded.new("a=b;c=d").split_html_form
    #  #=> [[#<Escape::PercentEncoded: a>, #<Escape::PercentEncoded: b>],
    #       [#<Escape::PercentEncoded: c>, #<Escape::PercentEncoded: d>]]
    #
    #  Escape::PercentEncoded.new("%3D=%3F").split_html_form
    #  #=> [[#<Escape::PercentEncoded: %3D>, #<Escape::PercentEncoded: %3F>]]
    #
    def split_html_form
      assoc = []
      @str.split(/[&;]/, -1).each {|s|
        raise InvalidHTMLForm, "invalid: #{@str}" unless /=/ =~ s
        assoc << [PercentEncoded.new_no_dup($`), PercentEncoded.new_no_dup($')]
      }
      assoc
    end
  end

  # Escape.percent_encoding escapes URI non-unreserved characters using percent-encoding.
  # It returns an instance of PercentEncoded.
  #
  # The unreserved characters are alphabet, digit, hyphen, dot, underscore and tilde.
  # [RFC 3986]
  #
  #  Escape.percent_encoding("foo") #=> #<Escape::PercentEncoded: foo>
  #
  #  Escape.percent_encoding(' !"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~')
  #  #=> #<Escape::PercentEncoded: %20%21%22%23%24%25%26%27%28%29%2A%2B%2C-.%2F%3A%3B%3C%3D%3E%3F%40%5B%5C%5D%5E_%60%7B%7C%7D~>
  def percent_encoding(str)
    s = str.gsub(%r{[^a-zA-Za-z0-9\-._~]}n) {
      '%' + $&.unpack("H2")[0].upcase
    }
    PercentEncoded.new_no_dup(s)
  end

  # Escape.uri_segment escapes URI segment using percent-encoding.
  # It returns an instance of PercentEncoded.
  #
  #  Escape.uri_segment("a/b") #=> #<Escape::PercentEncoded: a%2Fb>
  #
  # The segment is "/"-splitted element after authority before query in URI, as follows.
  #
  #   scheme://authority/segment1/segment2/.../segmentN?query#fragment
  #
  # See RFC 3986 for details of URI.
  def uri_segment(str)
    # pchar - pct-encoded = unreserved / sub-delims / ":" / "@"
    # unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
    # sub-delims = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
    s = str.gsub(%r{[^A-Za-z0-9\-._~!$&'()*+,;=:@]}n) {
      '%' + $&.unpack("H2")[0].upcase
    }
    PercentEncoded.new_no_dup(s)
  end

  # Escape.uri_path escapes URI path using percent-encoding.
  #
  # The given path should be one of follows.
  # * a sequence of (non-escaped) segments separated by "/".  (The segments cannot contains "/".)
  # * an array containing (non-escaped) segments.  (The segments may contains "/".)
  #
  # It returns an instance of PercentEncoded.
  #
  #  Escape.uri_path("a/b/c") #=> #<Escape::PercentEncoded: a/b/c>
  #  Escape.uri_path("a?b/c?d/e?f") #=> #<Escape::PercentEncoded: a%3Fb/c%3Fd/e%3Ff>
  #  Escape.uri_path(%w[/d f]) #=> "%2Fd/f"
  #
  # The path is the part after authority before query in URI, as follows.
  #
  #   scheme://authority/path#fragment
  #
  # See RFC 3986 for details of URI.
  #
  # Note that this function is not appropriate to convert OS path to URI.
  def uri_path(arg)
    if arg.respond_to? :to_ary
      s = arg.map {|elt| uri_segment(elt) }.join('/')
    else
      s = arg.gsub(%r{[^/]+}n) { uri_segment($&) }
    end
    PercentEncoded.new_no_dup(s)
  end

  # :stopdoc:
  def html_form_fast(pairs, sep='&')
    s = pairs.map {|k, v|
      # query-chars - pct-encoded - x-www-form-urlencoded-delimiters =
      #   unreserved / "!" / "$" / "'" / "(" / ")" / "*" / "," / ":" / "@" / "/" / "?"
      # query-char - pct-encoded = unreserved / sub-delims / ":" / "@" / "/" / "?"
      # query-char = pchar / "/" / "?" = unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?"
      # unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
      # sub-delims = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
      # x-www-form-urlencoded-delimiters = "&" / "+" / ";" / "="
      k = k.gsub(%r{[^0-9A-Za-z\-\._~:/?@!\$'()*,]}n) {
        '%' + $&.unpack("H2")[0].upcase
      }
      v = v.gsub(%r{[^0-9A-Za-z\-\._~:/?@!\$'()*,]}n) {
        '%' + $&.unpack("H2")[0].upcase
      }
      "#{k}=#{v}"
    }.join(sep)
    PercentEncoded.new_no_dup(s)
  end
  # :startdoc:

  # Escape.html_form composes HTML form key-value pairs as a x-www-form-urlencoded encoded string.
  # It returns an instance of PercentEncoded.
  #
  # Escape.html_form takes an array of pair of strings or
  # an hash from string to string.
  #
  #  Escape.html_form([["a","b"], ["c","d"]]) #=> #<Escape::PercentEncoded: a=b&c=d>
  #  Escape.html_form({"a"=>"b", "c"=>"d"}) #=> #<Escape::PercentEncoded: a=b&c=d>
  #
  # In the array form, it is possible to use same key more than once.
  # (It is required for a HTML form which contains
  # checkboxes and select element with multiple attribute.)
  #
  #  Escape.html_form([["k","1"], ["k","2"]]) #=> #<Escape::PercentEncoded: k=1&k=2>
  #
  # If the strings contains characters which must be escaped in x-www-form-urlencoded,
  # they are escaped using %-encoding.
  #
  #  Escape.html_form([["k=","&;="]]) #=> #<Escape::PercentEncoded: k%3D=%26%3B%3D>
  #
  # The separator can be specified by the optional second argument.
  #
  #  Escape.html_form([["a","b"], ["c","d"]], ";") #=> #<Escape::PercentEncoded: a=b;c=d>
  #
  # See HTML 4.01 for details.
  def html_form(pairs, sep='&')
    r = ''
    first = true
    pairs.each {|k, v|
      # query-chars - pct-encoded - x-www-form-urlencoded-delimiters =
      #   unreserved / "!" / "$" / "'" / "(" / ")" / "*" / "," / ":" / "@" / "/" / "?"
      # query-char - pct-encoded = unreserved / sub-delims / ":" / "@" / "/" / "?"
      # query-char = pchar / "/" / "?" = unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?"
      # unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
      # sub-delims = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
      # x-www-form-urlencoded-delimiters = "&" / "+" / ";" / "="
      r << sep if !first
      first = false
      k.each_byte {|byte|
        ch = byte.chr
        if %r{[^0-9A-Za-z\-\._~:/?@!\$'()*,]}n =~ ch
          r << "%" << ch.unpack("H2")[0].upcase
        else
          r << ch
        end
      }
      r << '='
      v.each_byte {|byte|
        ch = byte.chr
        if %r{[^0-9A-Za-z\-\._~:/?@!\$'()*,]}n =~ ch
          r << "%" << ch.unpack("H2")[0].upcase
        else
          r << ch
        end
      }
    }
    PercentEncoded.new_no_dup(r)
  end

  class HTMLEscaped < StringWrapper
  end

  # :stopdoc:
  HTML_TEXT_ESCAPE_HASH = {
    '&' => '&amp;',
    '<' => '&lt;',
    '>' => '&gt;',
  }
  # :startdoc:

  # Escape.html_text escapes a string appropriate for HTML text using character references.
  # It returns an instance of HTMLEscaped.
  #
  # It escapes 3 characters:
  # * '&' to '&amp;'
  # * '<' to '&lt;'
  # * '>' to '&gt;'
  #
  #  Escape.html_text("abc") #=> #<Escape::HTMLEscaped: abc>
  #  Escape.html_text("a & b < c > d") #=> #<Escape::HTMLEscaped: a &amp; b &lt; c &gt; d>
  #
  # This function is not appropriate for escaping HTML element attribute
  # because quotes are not escaped.
  def html_text(str)
    s = str.gsub(/[&<>]/) {|ch| HTML_TEXT_ESCAPE_HASH[ch] }
    HTMLEscaped.new_no_dup(s)
  end

  # :stopdoc:
  HTML_ATTR_ESCAPE_HASH = {
    '&' => '&amp;',
    '<' => '&lt;',
    '>' => '&gt;',
    '"' => '&quot;',
  }
  # :startdoc:

  class HTMLAttrValue < StringWrapper
  end

  # Escape.html_attr_value encodes a string as a double-quoted HTML attribute using character references.
  # It returns an instance of HTMLAttrValue.
  #
  #  Escape.html_attr_value("abc") #=> #<Escape::HTMLAttrValue: "abc">
  #  Escape.html_attr_value("a&b") #=> #<Escape::HTMLAttrValue: "a&amp;b">
  #  Escape.html_attr_value("ab&<>\"c") #=> #<Escape::HTMLAttrValue: "ab&amp;&lt;&gt;&quot;c">
  #  Escape.html_attr_value("a'c") #=> #<Escape::HTMLAttrValue: "a'c">
  #
  # It escapes 4 characters:
  # * '&' to '&amp;'
  # * '<' to '&lt;'
  # * '>' to '&gt;'
  # * '"' to '&quot;'
  #
  def html_attr_value(str)
    s = '"' + str.gsub(/[&<>"]/) {|ch| HTML_ATTR_ESCAPE_HASH[ch] } + '"'
    HTMLAttrValue.new_no_dup(s)
  end

  # MIMEParameter represents parameter, token, quoted-string in MIME.
  # parameter and token is defined in RFC 2045.
  # quoted-string is defined in RFC 822.
  class MIMEParameter < StringWrapper
  end

  # predicate for MIME token.
  #
  # token is a sequence of any (US-ASCII) CHAR except SPACE, CTLs, or tspecials.
  def mime_token?(str)
    /\A[!\#-'*-+\-.0-9A-Z^-~]+\z/ =~ str ? true : false
  end

  # Escape.rfc2822_quoted_string escapes a string as quoted-string defined in RFC 2822.
  # It returns an instance of MIMEParameter.
  #
  # It cannot represents CR, LF and NUL. 
  # If the string contained them, ArgumentError is raised.
  #
  def rfc2822_quoted_string(str)
    if /[\r\n\0]/ =~ str
      raise ArgumentError, "CR, LF or NUL contained: #{str.inspect}"
    end
    s = '"' + str.gsub(/["\\]/, '\\\\\&') + '"'
    MIMEParameter.new_no_dup(s)
  end

  # Escape.mime_parameter_value escapes a string as MIME parameter value in RFC 2045.
  # It returns an instance of MIMEParameter.
  #
  # MIME parameter value is token or quoted-string.
  # token is returned if possible.
  def mime_parameter_value(str)
    if mime_token?(str)
      MIMEParameter.new(str)
    else
      rfc2822_quoted_string(str)
    end
  end

end
